File: code/mobile/lib/features/community/domain/community_message.dart
enum CommunityMessageType { user, memberJoined, memberLeft }

class CommunityMessage {
  final int id;
  final String? clientMessageId;
  final int communityId;
  final int authorId;
  final String content;
  final CommunityMessageType type;
  final int? parentMessageId;
  final List<String> tags;
  final List<String> images;
  final String? gifUrl;
  final bool hasSpoiler;
  final int heartCount;
  final bool deleted;
  final DateTime createdAt;
  final DateTime? editedAt;

  const CommunityMessage({
    required this.id,
    this.clientMessageId,
    required this.communityId,
    required this.authorId,
    required this.content,
    required this.type,
    this.parentMessageId,
    this.tags = const [],
    required this.images,
    this.gifUrl,
    required this.hasSpoiler,
    this.heartCount = 0,
    required this.deleted,
    required this.createdAt,
    this.editedAt,
  });

  CommunityMessage copyWith({
    int? id,
    String? clientMessageId,
    int? communityId,
    int? authorId,
    String? content,
    CommunityMessageType? type,
    int? parentMessageId,
    List<String>? tags,
    List<String>? images,
    String? gifUrl,
    bool? hasSpoiler,
    int? heartCount,
    bool? deleted,
    DateTime? createdAt,
    DateTime? editedAt,
  }) {
    return CommunityMessage(
      id: id ?? this.id,
      clientMessageId: clientMessageId ?? this.clientMessageId,
      communityId: communityId ?? this.communityId,
      authorId: authorId ?? this.authorId,
      content: content ?? this.content,
      type: type ?? this.type,
      parentMessageId: parentMessageId ?? this.parentMessageId,
      tags: tags ?? this.tags,
      images: images ?? this.images,
      gifUrl: gifUrl ?? this.gifUrl,
      hasSpoiler: hasSpoiler ?? this.hasSpoiler,
      heartCount: heartCount ?? this.heartCount,
      deleted: deleted ?? this.deleted,
      createdAt: createdAt ?? this.createdAt,
      editedAt: editedAt ?? this.editedAt,
    );
  }
}

CommunityMessageType parseCommunityMessageType(String? raw) {
  final normalized = (raw ?? '').trim().toUpperCase();
  if (normalized.contains('MEMBER_JOINED')) {
    return CommunityMessageType.memberJoined;
  }
  if (normalized.contains('MEMBER_LEFT')) {
    return CommunityMessageType.memberLeft;
  }
  return CommunityMessageType.user;
}

CommunityMessageType resolveCommunityMessageType(
  String? raw, {
  String? content,
  List<String>? images,
  String? gifUrl,
  List<String>? tags,
  int? parentMessageId,
  bool? deleted,
}) {
  final parsed = parseCommunityMessageType(raw);
  if (parsed == CommunityMessageType.user) {
    return parsed;
  }

  final hasMedia =
      (images != null && images.isNotEmpty) ||
      (gifUrl != null && gifUrl.isNotEmpty);
  final hasText = content != null && content.trim().isNotEmpty;
  final hasTags = tags != null && tags.isNotEmpty;
  final hasReply = parentMessageId != null;
  final isDeleted = deleted == true;

  if (hasText || hasMedia || hasTags || hasReply || isDeleted) {
    return CommunityMessageType.user;
  }

  return parsed;
}

String serializeCommunityMessageType(CommunityMessageType type) {
  switch (type) {
    case CommunityMessageType.memberJoined:
      return 'MEMBER_JOINED';
    case CommunityMessageType.memberLeft:
      return 'MEMBER_LEFT';
    case CommunityMessageType.user:
    default:
      return 'USER';
  }
}

File: code/back/src/main/java/com/biblioo/community/domain/model/CommunityMessage.java
package com.biblioo.community.domain.model;

import jakarta.persistence.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import lombok.*;

@Entity
@Table(
    name = "community_messages",
    indexes = {
      @Index(name = "idx_cmsg_community_id", columnList = "community_id, id DESC"),
      @Index(name = "idx_cmsg_parent", columnList = "parent_message_id"),
      @Index(name = "idx_cmsg_author", columnList = "author_id")
    })
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class CommunityMessage {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  private Long id;

  @Column(name = "community_id", nullable = false)
  private Long communityId;

  @Column(name = "author_id", nullable = false)
  private Long authorId;

  @Column(nullable = true, length = 4000)
  private String content;

  @Column(name = "parent_message_id")
  private Long parentMessageId;

  @ElementCollection(fetch = FetchType.EAGER)
  @CollectionTable(name = "community_message_tags", joinColumns = @JoinColumn(name = "message_id"))
  @Column(name = "tag", length = 50)
  @Builder.Default
  private Set<String> tags = new HashSet<>();

  @ElementCollection(fetch = FetchType.EAGER)
  @CollectionTable(
      name = "community_message_images",
      joinColumns = @JoinColumn(name = "message_id"))
  @Column(name = "image_url", length = 2048)
  @Builder.Default
  private List<String> images = new ArrayList<>();

  @Column(name = "gif_url", length = 2048)
  private String gifUrl;

  @Column(name = "has_spoiler", nullable = false)
  @Builder.Default
  private boolean hasSpoiler = false;

  @Enumerated(EnumType.STRING)
  @Column(nullable = false, length = 20)
  @Builder.Default
  private MessageType type = MessageType.USER;

  @Column(name = "client_message_id", length = 36)
  private String clientMessageId;

  @Column(name = "heart_count", nullable = false)
  @Builder.Default
  private int heartCount = 0;

  @Column(name = "is_deleted", nullable = false)
  @Builder.Default
  private boolean deleted = false;

  @Column(name = "created_at", nullable = false, updatable = false)
  private LocalDateTime createdAt;

  @Column(name = "edited_at")
  private LocalDateTime editedAt;

  @Version private Long version;

  @PrePersist
  protected void onCreate() {
    this.createdAt = LocalDateTime.now();
  }
}
